File: controller/blackscholes.go
package controller

import (
	"math"
	"errors"
	"gonum.org/v1/gonum/stat/distuv"
)

// GetProbabilityFactors - Compute d1 and d2 in Black Scholes
// Formula:
// 	d1 = (log(S/K) + (r + sigma^2/2*tau)) / (sigma*sqrt(tau))
// 	tau = T - t = time to maturity
// 	r = rate of return
// 	S = spot price
// 	K = strike price
// 	sigma = implied volatility
// Arguments:
// 	spot - Spot price
// 	strike - Strike price
// 	sigma - Annualized implied volatility
// 	tau - Time to expiry (annualized)
// 	rate - The risk-free rate
// Returns:
// 	d1 Probability factor one 
// 	d2 Probability factor one 
func GetProbabilityFactors(
	spot float64,
	strike float64,
	sigma float64,
	tau float64,
	rate float64,
) (float64, float64) {
	logRatio := math.Log(spot / strike)
	vol := sigma * math.Sqrt(tau)
	d1 := (logRatio + (rate + tau * math.Pow(sigma, 2) / 2)) / vol
	d2 := d1 - vol
	return d1, d2
}

// GetMarkPrice - Helper function to return call or put prices
// Arguments: 
// 	See `GetCallPrice` and `GetPutPrice`
// 	isCall - Boolean that is true if call else put
func GetMarkPrice(
	spot float64,
	strike float64,
	sigma float64,
	tau float64,
	rate float64,
	isCall bool,
) (float64) {
	var price float64
	if (isCall) {
		price = GetCallPrice(spot, strike, sigma, tau, rate)
	} else {
		price = GetPutPrice(spot, strike, sigma, tau, rate)
	}
	return price
}

// GetCallPrice - Compute Black Scholes price of call
// Formula:
// 	C = SN(d1)-Ke^{-rt}N(d2)
func GetCallPrice(
	spot float64,
	strike float64,
	sigma float64,
	tau float64,
	rate float64,
) (float64) {
	d1, d2 := GetProbabilityFactors(spot, strike, sigma, tau, rate)
	expTerm := math.Exp(-rate * tau)
	price := spot * distuv.UnitNormal.CDF(d1) - strike * expTerm * distuv.UnitNormal.CDF(d2)
	return price
}

// GetPutPrice - Compute Black Scholes price of put
// Formula:
// 	P = Ke^{-rt}N(-d2)-SN(-d1)
func GetPutPrice(
	spot float64,
	strike float64,
	sigma float64,
	tau float64,
	rate float64,
) (float64) {
	d1, d2 := GetProbabilityFactors(spot, strike, sigma, tau, rate)
	expTerm := math.Exp(-rate * tau)
	price := strike * expTerm * distuv.UnitNormal.CDF(-d2) - spot * distuv.UnitNormal.CDF(-d1)
	return price
}

// GetSigmaByBisection - Solve for volatility from call price iteratively using Bisection method
// Resources:
// 	https://en.wikipedia.org/wiki/Bisection_method
// Notes:
// 	We found this to be more stable and efficient than Newton Raphson
// 	which struggles with Vega -> 0 for strikes from the spot
// Arguments:
// 	isCall - boolean that is true if option is a call
// 	maxIter - Maximum number of iterations for convergence
// 	tolerance - Maximum discrepancy to define convergence
func GetSigmaByBisection(
	spot float64,
	strike float64,
	tau float64,
	rate float64,
	price float64,
	isCall bool,
	maxIter int,
	tolerance float64,
) (float64, error) {
	if (price > strike) {
		return 0, errors.New("GetSigmaByBisection: price must be < strike")
	}

	// Initialize left and right bounds
	var leftSigma float64 = 0.001
	var rightSigma float64 = 10
	var midSigma float64 = (leftSigma + rightSigma) / 2


	for i := 0; i < maxIter; i++ {
		// Compute Black Scholes price
		leftPrice := GetMarkPrice(spot, strike, leftSigma, tau, rate, isCall)
		midPrice := GetMarkPrice(spot, strike, midSigma, tau, rate, isCall)

		// Calculate difference to trade price
		leftDiff := leftPrice - price
		midDiff := midPrice - price

		if math.Abs(midDiff) < tolerance {
			break;
		}

		// Check if the signs are the same
		if (midDiff >= 0) == (leftDiff >= 0) {
			leftSigma = midSigma;
		} else {
			rightSigma = midSigma;
		}

		// Update mid point
		midSigma = (leftSigma + rightSigma) / 2
	}

	if (midSigma <= 0) {
		return 0, errors.New("GetSigmaByBisection: sigma must be > 0")
	}
	return midSigma, nil
}

File: scripts/utils/blackscholes.py
import math
import numpy as np
from scipy.stats import norm


def get_probability_factors(spot, strike, sigma, tau, rate):
  d1 = (math.log(spot/strike)+(rate+tau*sigma**2/2.0))/(sigma*math.sqrt(tau))
  d2 = d1-sigma*math.sqrt(tau)
  return d1, d2


def get_call_price(spot, strike, sigma, tau, rate):
  d1, d2 = get_probability_factors(spot, strike, sigma, tau, rate)
  price = spot*norm.cdf(d1)-strike*math.exp(-rate*tau)*norm.cdf(d2)
  return price


def get_put_price(spot, strike, sigma, tau, rate):
  d1, d2 = get_probability_factors(spot, strike, sigma, tau, rate)
  price = strike*math.exp(-rate*tau)*norm.cdf(-d2)-spot*norm.cdf(-d1)
  return price


def get_vega(spot, strike, sigma, tau, rate):
  d1, _ = get_probability_factors(spot, strike, sigma, tau, rate)
  vega = spot*math.sqrt(tau)*norm.pdf(d1)
  return vega


def brute_force_iv(spot, strike, rate, tau, market, is_call=True):
  candidates = np.arange(0.0001,4,0.001)
  diffs = np.zeros_like(candidates)

  for i in range(len(candidates)):
    candidate = candidates[i]
    if is_call:
      diff = get_call_price(spot, strike, candidate, tau, rate) - market
    else:
      diff = get_put_price(spot, strike, candidate, tau, rate) - market
    diffs[i] = abs(diff)
  
  best_i = np.argmin(diffs)
  return candidates[best_i], diffs[best_i]
